File: export.py
#!/usr/bin/env python3
"""Render .blend assets into per-direction PNGs plus a JSON manifest.

For each *.blend file found, invokes Blender headlessly (blender_scene_export.py)
to render its four standard isometric directions — NE/NW/SE/SW, auto-framed to
the object's bounding box unless the scene already has a camera by that name —
plus any other camera already in the scene, and writes:

  <out>/<stem>_<direction-or-camera>.png   one per rendered view
  <out>/<stem>.json                        manifest describing them all

See blender_scene_export.py for the camera/anchor conventions, and README.md
for the JSON schema.

Usage:
  python3 export.py <blend-file-or-directory> [--out DIR] [--force]
                     [--px-per-unit 32] [--supersample 4] [--pitch 35.264]
"""

import argparse
import json
import shutil
import subprocess
import sys
from pathlib import Path

SCRIPT_DIR = Path(__file__).parent
BLENDER_SCRIPT = SCRIPT_DIR / "blender_scene_export.py"


def die(msg):
    print(f"error: {msg}", file=sys.stderr)
    sys.exit(1)


def find_blends(path: Path) -> list[Path]:
    if path.is_file():
        return [path]
    return sorted(path.rglob("*.blend"))


def stale(blend: Path, out_dir: Path) -> bool:
    manifest = out_dir / f"{blend.stem}.json"
    if not manifest.exists():
        return True
    return blend.stat().st_mtime > manifest.stat().st_mtime


def render_blend(blend: Path, out_dir: Path, px_per_unit: float, png_px_per_unit: float, pitch_deg: float):
    out_json = out_dir / f"{blend.stem}.json"
    result = subprocess.run(
        ["blender", str(blend), "--background", "--python", str(BLENDER_SCRIPT), "--",
         str(out_dir), str(px_per_unit), str(png_px_per_unit), str(pitch_deg)],
        capture_output=True, text=True,
    )
    if result.returncode != 0 or not out_json.exists():
        print(f"  ✗ {blend.stem} failed:\n{result.stderr.strip()}", file=sys.stderr)
        return None
    with out_json.open() as f:
        return json.load(f)


def process(blend: Path, out_dir: Path, args) -> bool:
    print(f"{blend.name}:")
    data = render_blend(blend, out_dir, args.px_per_unit, args.px_per_unit * args.supersample, args.pitch)
    if data is None:
        return False
    manifest = out_dir / f"{blend.stem}.json"
    views = data.get("views", {})
    n_dirs = sum(1 for key in views if key.lower() in ("ne", "nw", "se", "sw"))
    n_extra = len(views) - n_dirs
    n_anchors = len(next(iter(views.values()), {}).get("anchors", {}))
    print(f"  wrote {manifest.name} ({n_dirs} direction(s), {n_extra} extra camera(s), {n_anchors} anchor(s))")
    return True


def main():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("path", type=Path, help="a .blend file, or a directory to scan recursively")
    parser.add_argument("--out", type=Path, default=None, help="output directory (default: alongside each .blend)")
    parser.add_argument("--force", action="store_true", help="re-render even if the manifest is up to date")
    parser.add_argument("--px-per-unit", type=float, default=32.0,
                         help="game display pixels per Blender unit (default: 32)")
    parser.add_argument("--supersample", type=float, default=4.0,
                         help="rendered PNG pixels per display pixel, for auto-framed cameras (default: 4)")
    parser.add_argument("--pitch", type=float, default=35.264,
                         help="camera pitch below horizontal in degrees, for auto-framed cameras "
                              "(default: true isometric, 35.264)")
    args = parser.parse_args()

    if shutil.which("blender") is None:
        die("blender not found — install from https://www.blender.org/download/ "
            "(or `brew install --cask blender`) and ensure it's on PATH")

    blends = find_blends(args.path)
    if not blends:
        die(f"no .blend files found under {args.path}")

    ok = 0
    for blend in blends:
        out_dir = args.out if args.out else blend.parent
        out_dir.mkdir(parents=True, exist_ok=True)
        if not args.force and not stale(blend, out_dir):
            print(f"{blend.name}: up to date, skipping")
            ok += 1
            continue
        if process(blend, out_dir, args):
            ok += 1

    print(f"\n{ok}/{len(blends)} asset(s) exported.")
    if ok < len(blends):
        sys.exit(1)


if __name__ == "__main__":
    main()

File: cmd/export.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"

	"github.com/spf13/cobra"

	"github.com/TheWozard/blexport/internal/blender"
	"github.com/TheWozard/blexport/internal/manifest"
)

var (
	exportOut         string
	exportForce       bool
	exportPxPerUnit   float64
	exportSupersample float64
	exportPitch       float64
)

var exportCmd = &cobra.Command{
	Use:   "export <blend-file-or-directory>",
	Short: "Render every stale .blend file under path",
	Args:  cobra.ExactArgs(1),
	RunE:  runExport,
}

func init() {
	exportCmd.Flags().StringVar(&exportOut, "out", "", "output directory (default: alongside each .blend)")
	exportCmd.Flags().BoolVar(&exportForce, "force", false, "re-render even if the manifest is up to date")
	exportCmd.Flags().Float64Var(&exportPxPerUnit, "px-per-unit", 32, "game display pixels per Blender unit")
	exportCmd.Flags().Float64Var(&exportSupersample, "supersample", 4, "rendered PNG pixels per display pixel, for auto-framed cameras")
	exportCmd.Flags().Float64Var(&exportPitch, "pitch", 35.264, "camera pitch below horizontal, in degrees, for auto-framed cameras")
	rootCmd.AddCommand(exportCmd)
}

func runExport(cmd *cobra.Command, args []string) error {
	cmd.SilenceUsage = true

	if _, err := exec.LookPath("blender"); err != nil {
		return fmt.Errorf("blender not found on PATH — install from https://www.blender.org/download/ (or `brew install --cask blender`)")
	}

	blends, err := findBlends(args[0])
	if err != nil {
		return err
	}
	if len(blends) == 0 {
		return fmt.Errorf("no .blend files found under %s", args[0])
	}

	ok := 0
	for _, blend := range blends {
		outDir := exportOut
		if outDir == "" {
			outDir = filepath.Dir(blend)
		}
		if err := os.MkdirAll(outDir, 0755); err != nil {
			return err
		}

		stem := strings.TrimSuffix(filepath.Base(blend), filepath.Ext(blend))
		manifestPath := filepath.Join(outDir, stem+".json")

		if !exportForce {
			isStale, err := manifest.Stale(blend, manifestPath)
			if err != nil {
				return err
			}
			if !isStale {
				fmt.Printf("%s: up to date, skipping\n", filepath.Base(blend))
				ok++
				continue
			}
		}

		if processBlend(blend, outDir, manifestPath) {
			ok++
		}
	}

	fmt.Printf("\n%d/%d asset(s) exported.\n", ok, len(blends))
	if ok < len(blends) {
		os.Exit(1)
	}
	return nil
}

// processBlend renders one .blend and prints a per-asset summary. It
// reports success/failure itself (rather than returning an error) so one
// bad asset doesn't abort the batch, matching export.py's behavior.
func processBlend(blend, outDir, manifestPath string) bool {
	fmt.Printf("%s:\n", filepath.Base(blend))

	pngPxPerUnit := exportPxPerUnit * exportSupersample
	if err := blender.Run(blend, outDir, exportPxPerUnit, pngPxPerUnit, exportPitch); err != nil {
		fmt.Fprintf(os.Stderr, "  ✗ %s\n", err)
		return false
	}

	data, err := os.ReadFile(manifestPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "  ✗ blender exited ok but wrote no manifest: %s\n", err)
		return false
	}
	var m manifest.Manifest
	if err := json.Unmarshal(data, &m); err != nil {
		fmt.Fprintf(os.Stderr, "  ✗ parsing manifest: %s\n", err)
		return false
	}

	assets := m.AllAssets()
	fmt.Printf("  wrote %s (%d asset(s)):\n", filepath.Base(manifestPath), len(assets))
	for name, asset := range assets {
		fmt.Printf("    %s: %s\n", name, summarizeViews(asset.Views))
	}
	return true
}

func summarizeViews(views map[string]manifest.View) string {
	compassDirs := map[string]bool{"ne": true, "nw": true, "se": true, "sw": true}
	nDirs, nAnchors := 0, 0
	for key, v := range views {
		if compassDirs[strings.ToLower(key)] {
			nDirs++
		}
		if nAnchors == 0 {
			nAnchors = len(v.Anchors)
		}
	}
	return fmt.Sprintf("%d direction(s), %d extra camera(s), %d anchor(s)", nDirs, len(views)-nDirs, nAnchors)
}

// findBlends returns path itself if it's a .blend file, or every .blend
// found recursively under it if it's a directory, sorted for stable output.
func findBlends(path string) ([]string, error) {
	info, err := os.Stat(path)
	if err != nil {
		return nil, fmt.Errorf("accessing %s: %w", path, err)
	}
	if !info.IsDir() {
		return []string{path}, nil
	}

	var blends []string
	err = filepath.WalkDir(path, func(p string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if !d.IsDir() && strings.EqualFold(filepath.Ext(p), ".blend") {
			blends = append(blends, p)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Strings(blends)
	return blends, nil
}
